morethuente.go

```diff
@@ -0,0 +1,384 @@
+
+	mt.x, mt.fx, mt.gx = 0, f, g
+	mt.y, mt.fy, mt.gy = 0, f, g
+
+	mt.lower = 0
+	mt.upper = step + mtMaxGrowthFactor*step
+
+	mt.width[0] = mt.MaximumStep - mt.MinimumStep
+	mt.width[1] = 2 * mt.width[0]
+
+	mt.step = step
+	return FuncEvaluation | GradEvaluation
+}
+
+func (mt *MoreThuente) Iterate(f, g float64) (Operation, float64, error) {
+	if mt.stage == 0 {
+		panic("morethuente: Init has not been called")
+	}
+
+	gTest := mt.DecreaseFactor * mt.gInit
+	fTest := mt.fInit + mt.step*gTest
+
+	if mt.bracketed {
+		if mt.step <= mt.lower || mt.step >= mt.upper || mt.upper-mt.lower <= mt.StepTolerance*mt.upper {
+			// step contains the best step found (see below).
+			return NoOperation, mt.step, ErrLinesearcherFailure
+		}
+	}
+	if mt.step == mt.MaximumStep && f <= fTest && g <= gTest {
+		return NoOperation, mt.step, ErrLinesearcherBound
+	}
+	if mt.step == mt.MinimumStep && (f > fTest || g >= gTest) {
+		return NoOperation, mt.step, ErrLinesearcherFailure
+	}
+
+	// Test for convergence.
+	if f <= fTest && math.Abs(g) <= mt.CurvatureFactor*(-mt.gInit) {
+		return MajorIteration, mt.step, nil
+	}
+
+	if mt.stage == 1 && f <= fTest && g >= 0 {
+		mt.stage = 2
+	}
+
+	if mt.stage == 1 && f <= mt.fx && f > fTest {
+		// Lower function value but the decrease is not sufficient .
+
+		// Compute values and derivatives of the modified function at step, x, y.
+		fm := f - mt.step*gTest
+		fxm := mt.fx - mt.x*gTest
+		fym := mt.fy - mt.y*gTest
+		gm := g - gTest
+		gxm := mt.gx - gTest
+		gym := mt.gy - gTest
+		// Update x, y and step.
+		mt.nextStep(fxm, gxm, fym, gym, fm, gm)
+		// Recover values and derivates of the non-modified function at x and y.
+		mt.fx = fxm + mt.x*gTest
+		mt.fy = fym + mt.y*gTest
+		mt.gx = gxm + gTest
+		mt.gy = gym + gTest
+	} else {
+		// Update x, y and step.
+		mt.nextStep(mt.fx, mt.gx, mt.fy, mt.gy, f, g)
+	}
+
+	if mt.bracketed {
+		// Monitor the length of the bracketing interval. If the interval has
+		// not been reduced sufficiently after two steps, use bisection to
+		// force its length to zero.
+		width := mt.y - mt.x
+		if math.Abs(width) >= 0.66*mt.width[1] {
+			mt.step = mt.x + 0.5*width
+		}
+		mt.width[0], mt.width[1] = math.Abs(width), mt.width[0]
+	}
+
+	if mt.bracketed {
+		mt.lower = math.Min(mt.x, mt.y)
+		mt.upper = math.Max(mt.x, mt.y)
+	} else {
+		mt.lower = mt.step + mtMinGrowthFactor*(mt.step-mt.x)
+		mt.upper = mt.step + mtMaxGrowthFactor*(mt.step-mt.x)
+	}
+
+	// Force the step to be within MininmumStep and MaximumStep.
+	mt.step = math.Max(mt.MinimumStep, math.Min(mt.step, mt.MaximumStep))
+
+	if mt.bracketed {
+		if mt.step <= mt.lower || mt.step >= mt.upper || mt.upper-mt.lower <= mt.StepTolerance*mt.upper {
+			// If further progress is not possible, set step to the best step
+			// obtained during the search.
+			mt.step = mt.x
+		}
+	}
+
+	return FuncEvaluation | GradEvaluation, mt.step, nil
+}
+
+// nextStep computes the next safeguarded step and updates the interval that
```

Comment:
I tried to factor out the interpolation code but each case differs a bit in such a way that I gave up. Still I feel that it should be possible to make it nicer, so it would be certainly nice if you could check it.